### DISCUSSION TITLE
openssl 3.0 & 3.0.7 - how to enable legacy provider?

### DISCUSSION THREAD

**Discussion by prasad1240** [openssl/openssl#21665] [Q&A] [answered]
Hi Team,

Downloaded latest openssl 3.0 & 3.0.7 code and executed below command in mac 

**Command:-**    
 ./configure ios64-xcrun no-tests -no-shared enable-weak-ssl-ciphers

**Added code in our project**
      
            int legacyValue = OSSL_PROVIDER_available(NULL, "legacy");
           OSSL_PROVIDER *legacy_new = OSSL_PROVIDER_try_load(NULL, "legacy", 1);
    OSSL_PROVIDER *defaultProvider_new = OSSL_PROVIDER_try_load(NULL, "default", 1);

        /* Load Multiple providers into the default (NULL) library context */
           legacy = OSSL_PROVIDER_load(NULL, "legacy");
           if (legacy == NULL) {
               printf("Failed to load Legacy provider\n");
           }

Now legacy provider is not loading.

**Could you please help us?**
**Can someone please review my changes and see what might still be missing?**

**Command:-** perl configdata.pm --dump

`
Command line (with current working directory = .):

    perl ./Configure ios64-xcrun no-tests -no-shared --openssldir=/usr/local/ssl OPENSSL_CONF=./openssl.cnf enable-weak-ssl-ciphers

Perl information
    perl
    5.30.3 for darwin-thread-multi-2level
Enabled features
    aria
    asm
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    ec
    ec2m
    ecdh
    ecdsa
    err
    filenames
    gost
    idea
    legacy
    makedepend
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    pic
    pinshared
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    threads
    tls
    ts
    ui-console
    weak-ssl-ciphers
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method
Disabled features
    acvp-tests          [cascade]        OPENSSL_NO_ACVP_TESTS
    afalgeng            [cascade]        OPENSSL_NO_AFALGENG
    asan                [default]        OPENSSL_NO_ASAN
    async               [config]         OPENSSL_NO_ASYNC
    buildtest-c++       [default]        
    capieng             [cascade]        OPENSSL_NO_CAPIENG
    crypto-mdebug       [default]        OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]        OPENSSL_NO_DEVCRYPTOENG
    dynamic-engine      [cascade]        
    ec_nistp_64_gcc_128 [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                 [default]        OPENSSL_NO_EGD
    engine              [config]         OPENSSL_NO_ENGINE (skip engines, crypto/engine)
    external-tests      [default]        OPENSSL_NO_EXTERNAL_TESTS
    fips                [default]        
    fips-securitychecks [cascade]        OPENSSL_NO_FIPS_SECURITYCHECKS
    fuzz-afl            [default]        OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]        OPENSSL_NO_KTLS
    loadereng           [cascade]        OPENSSL_NO_LOADERENG
    md2                 [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]        OPENSSL_NO_MSAN
    padlockeng          [cascade]        OPENSSL_NO_PADLOCKENG
    rc5                 [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]        OPENSSL_NO_SCTP
    shared              [option]         
    tests               [option]         OPENSSL_NO_TESTS
    trace               [default]        OPENSSL_NO_TRACE
    ubsan               [default]        OPENSSL_NO_UBSAN
    unit-test           [default]        OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch] OPENSSL_NO_UPLINK
    zlib                [default]        
    zlib-dynamic        [default]        
    ssl3                [default]        OPENSSL_NO_SSL3
    ssl3-method         [default]        OPENSSL_NO_SSL3_METHOD
Config target attributes
    AR => "ar",
    ARFLAGS => "qc",
    CC => "xcrun -sdk iphoneos cc",
    CFLAGS => "-O3",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib -c",
    RC => "windres",
    asm_arch => "aarch64",
    bn_ops => "SIXTY_FOUR_BIT_LONG RC4_CHAR",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-arch arm64 -mios-version-min=7.0.0 -fno-common",
    cppflags => "-D_REENTRANT",
    defines => [ "OPENSSL_BUILDING_OPENSSL" ],
    disable => [ "engine", "async" ],
    dso_scheme => "dlfcn",
    enable => [  ],
    includes => [  ],
    lflags => "-Wl,-search_paths_first",
    lib_cflags => "",
    lib_cppflags => "",
    lib_defines => [  ],
    module_cflags => "-fPIC",
    module_cxxflags => undef,
    module_ldflags => "-bundle",
    perl_platform => "Unix",
    perlasm_scheme => "ios64",
    shared_cflag => "-fPIC",
    shared_defines => [  ],
    shared_extension => ".\$(SHLIB_VERSION_NUMBER).dylib",
    shared_ldflag => "-dynamiclib -current_version \$(SHLIB_VERSION_NUMBER) -compatibility_version \$(SHLIB_VERSION_NUMBER)",
    shared_rcflag => "",
    shared_sonameflag => "-install_name \$(libdir)/",
    shared_target => "darwin-shared",
    sys_id => "iOS",
    thread_defines => [  ],
    thread_scheme => "pthreads",
    unistd => "<unistd.h>",
Recorded environment
    AR = 
    ARFLAGS = 
    AS = 
    ASFLAGS = 
    BUILDFILE = 
    CC = 
    CFLAGS = 
    CPP = 
    CPPDEFINES = 
    CPPFLAGS = 
    CPPINCLUDES = 
    CROSS_COMPILE = 
    CXX = 
    CXXFLAGS = 
    HASHBANGPERL = 
    LD = 
    LDFLAGS = 
    LDLIBS = 
    MT = 
    MTFLAGS = 
    OPENSSL_LOCAL_CONFIG_DIR = 
    PERL = 
    RANLIB = 
    RC = 
    RCFLAGS = 
    RM = 
    WINDRES = 
    __CNF_CFLAGS = 
    __CNF_CPPDEFINES = 
    __CNF_CPPFLAGS = 
    __CNF_CPPINCLUDES = 
    __CNF_CXXFLAGS = 
    __CNF_LDFLAGS = 
    __CNF_LDLIBS = 
Makevars
    AR              = ar
    ARFLAGS         = qc
    CC              = xcrun -sdk iphoneos cc
    CFLAGS          = -O3
    CPPDEFINES      = 
    CPPFLAGS        = 
    CPPINCLUDES     = 
    CXXFLAGS        = 
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         = 
    LDLIBS          = 
    PERL            = perl
    RANLIB          = ranlib -c
    RC              = windres
    RCFLAGS         = 
`

Now legacy provider is not loading.

**Could you please help us?**
**Can someone please review my changes and see what might still be missing?**

**Comment by t8m** [comment]
`no-shared` will disable building shared libcrypto and libssl libraries. But the legacy provider would still be built as a shared module. So you would need the legacy shared module present on the installed system and the paths would have to be correct. You can use `no-module` to make the legacy provider built-in.

**Comment by prasad1240** [comment]
Tried below command
**Command:-**
./Configure ios64-xcrun no-tests -no-shared -no-module --openssldir=/usr/local/ssl OPENSSL_CONF=./openssl.cnf enable-weak-ssl-ciphers

Now 
**Error in xcode:-**
`duplicate symbol '_bn_mul_mont' in:
    /xxxxx/openSSL/lib/libcrypto.a(libcrypto-lib-armv8-mont.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-armv8-mont.o)
duplicate symbol '_bn_div_words' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-bn_asm.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-bn_asm.o)
duplicate symbol '_bn_sqr_words' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-bn_asm.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-bn_asm.o)
duplicate symbol '_bn_mul_words' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-bn_asm.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-bn_asm.o)
duplicate symbol '_bn_add_words' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-bn_asm.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-bn_asm.o)
duplicate symbol '_bn_mul_add_words' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-bn_asm.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-bn_asm.o)
duplicate symbol '_bn_sub_words' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-bn_asm.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-bn_asm.o)
duplicate symbol '_bn_sqr_comba8' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-bn_asm.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-bn_asm.o)
duplicate symbol '_DES_encrypt3' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-des_enc.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-des_enc.o)
duplicate symbol '_DES_decrypt3' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-des_enc.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-des_enc.o)
duplicate symbol '_DES_encrypt2' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-des_enc.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-des_enc.o)
duplicate symbol '_DES_encrypt1' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-des_enc.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-des_enc.o)
duplicate symbol '_fcrypt_body' in:
    /XXXXX/openSSL/lib/libcrypto.a(libcrypto-lib-fcrypt_b.o)
    /XXXXX/openSSL/lib/libcrypto.a(liblegacy-lib-fcrypt_b.o)
`

ld: 19 duplicate symbols for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
@t8m Please check this error

**Comment by t8m** [comment]
Could you please open a new issue with that?

**Comment by prasad1240** [comment]
@t8m Thanks for the quick response.
Sure, will create a new ticket for duplicate symbols error

Meanwhile, we removed "no-shared" &"no-module" and tried it 

**Command:-**
 ./Configure ios64-xcrun no-tests -no-autoload-config --openssldir=/usr/local/ssl OPENSSL_CONF=./openssl.cnf enable-weak-ssl-ciphers
 
 Still legacy provider is not loading..

**Comment by t8m** [comment]
As I said above - is the legacy provider shared library placed at place where it is expected by the libcrypto? Is there anything like strace on iOS? If so, can you run your application under it to find out what it is trying to do?

**Comment by prasad1240** [comment]
Created a new ticket for duplicate symbols issue
https://github.com/openssl/openssl/issues/20124

**Comment by prasad1240** [comment]
@t8m I had placed the legacy provider shared library in the same path as libcrypto.a but legacy provider is not loading. 
Please suggest the next step.

**Comment by t8m** [comment]
> @t8m I had placed the legacy provider shared library in the same path as libcrypto.a but legacy provider is not loading. Please suggest the next step.

The provider has to be in /usr/local/lib/ossl-modules (or look at the MODULESDIR in the Makefile created by the configure command).

**Comment by t8m** [comment]
`make install` should place it at the right place.

**Comment by prasad1240** [comment]
1. Download openssl 3.0.7 code  
2. Executed command :-   `./Configure ios64-xcrun no-tests -no-autoload-config --openssldir=/usr/local/ssl OPENSSL_CONF=./openssl.cnf enable-weak-ssl-ciphers`
3. Copied provider liblegacy.a  into "/usr/local/lib/ossl-modules" path
4. Executed command:-  `make install`

@t8m Still legacy provider is not loading.
Please help us

**Comment by t8m** [comment]
liblegacy.a is a static library, that is not the dynamically loadable legacy provider module.

What is in the /usr/local/lib/ossl-modules directory after make install?

**Comment by t8m** [comment]
Unfortunately I have no particular knowledge about iOS so I do not know if there are any tools like strace that can be used to debug things. I am afraid that without examining what exactly is happening when you call OSSL_PROVIDER_load() I would not be able to help you further.

**Comment by prasad1240** [comment]
After `make install`command executed, "legacy.dylib" file is there in  /usr/local/lib/ossl-modules directory

**Comment by t8m** [comment]
That is correct. Then the only reasons for the failure can be either OS blocking the load or the legacy.dylib produced being somehow broken.

Could you print the error stack with `ERR_print_errors()` to find out if there are any errors on the stack?

**Comment by prasad1240** [comment]
We are parsing the p12 cert with "RC2-40-CBC" cipher and getting below error 

this `ERR_print_errors_fp(stderr)` error stack trace

```
8005E70101000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(/usr/local/lib/ossl-modules/legacy.dylib): dlopen(/usr/local/lib/ossl-modules/legacy.dylib, 0x0002): tried: '/usr/lib/system/introspection/legacy.dylib' (no such file), '/usr/local/lib/ossl-modules/legacy.dylib' (no such file)
8005E70101000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:152:
8005E70101000000:error:07880025:common libcrypto routines:provider_init:reason(524325):crypto/provider_core.c:912:name=legacy
8005E70101000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(/usr/local/lib/ossl-modules/fips.dylib): dlopen(/usr/local/lib/ossl-modules/fips.dylib, 0x0002): tried: '/usr/lib/27/01/2023 1:22:06:905 PM|FLMCPreferenceService|removeUserAgent[20195]|841<Info>: Clearing user agent
system/introspection/fips.dylib' (no such file), '/usr/local/lib/ossl-modules/fips.dylib' (no such file)
8005E70101000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:152:
8005E70101000000:error:07880025:common libcrypto routines:provider_init:reason(524325):crypto/provider_core.c:912:name=fips
8005E70101000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(/usr/local/lib/ossl-modules/legacy.dylib): dlopen(/usr/local/lib/ossl-modules/legacy.dylib, 0x0002): tried: '/usr/lib/system/introspection/legacy.dylib' (no such file), '/usr/local/lib/ossl-modules/legacy.dylib' (no such file)
8005E70101000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:152:
8005E70101000000:error:07880025:common libcrypto routines:provider_init:reason(524325):crypto/provider_core.c:912:name=legacy
8005E70101000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:373:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()

```
@t8m  is there any other way to parse the p12 cert with "RC2-40-CBC" cipher(legacy cipher) ?

**Comment by t8m** [comment]
The '/usr/local/lib/ossl-modules/legacy.dylib' (no such file) seems to contradict your https://github.com/openssl/openssl/issues/20112#issuecomment-1403410014

**Comment by prasad1240** [comment]
My mac has "/usr/local/lib/ossl-modules/legacy.dylib" path with the correct files but we are testing in iphone device so those paths are not present in ios device.
CC: @t8m  Please suggest us.

**Comment by t8m** [comment]
There is not much I can suggest further. I do not know any details about iOS.

**Comment by prasad1240** [comment]
@t8m is there any way to change the search path for "legacy.dylib" in crypto when trying to load legacy providers?

**Comment by t8m** [comment]
Yes, there is a OSSL_PROVIDER_set_default_search_path() function.

**Comment by prasad1240** [comment]
Now updated `MODULESDIR=./` key in "Make File" and generated ios openssl library. 

Getting below errors while loading "legacy" provider 
```
8085240101000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:
8085240101000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:152:
8085240101000000:error:07880025:common libcrypto routines:provider_init:reason(524325):crypto/provider_core.c:912:name=legacy
8085240101000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:
8085240101000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:152:
8085240101000000:error:07880025:common libcrypto routines:provider_init:reason(524325):crypto/provider_core.c:912:name=fips
8085240101000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:
8085240101000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:152:
8085240101000000:error:07880025:common libcrypto routines:provider_init:reason(524325):crypto/provider_core.c:912:name=legacy
8085240101000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:373:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
```
CC: @t8m

**Comment by t8m** [comment]
This means the legacy.dylib was not loaded for some reason. Unfortunately this does not show why.

**Comment by khusboo7** [comment]
@prasad1240 Even I am facing the same issue. Did you find any solution?
What value are you getting for ` int legacyValue = OSSL_PROVIDER_available(NULL, "legacy");`?

**Comment by prasad1240** [comment, marked as answer]
@khusboo7
`./Configure ios64-xcrun no-tests -no-shared -no-module --openssldir=/usr/local/ssl OPENSSL_CONF=./openssl.cnf enable-weak-ssl-ciphers`
Using above command, I was generated openssl library and legacy provider is loading.

**Reply by khusboo7** [reply]
Thank you @prasad1240. The solution worked.

**Comment by gururaghavendraprasad** [comment]
Hi @prasad1240 and @t8m,

We are facing issue with RC2-40-CBC cipher from legacy on OpenSSL3.0.8.
We are building the legacy provider statically enabling STATIC_LEGACY and also could load the legacy provider using "legacy = OSSL_PROVIDER_load(legacy_ctx, "legacy");".

Our question is do you able to decrypt the PKCS#12 package with RC2_40_CBC using legacy provider?
Any suggestion to activate the RC2_40_CBC from legacy or default provider is highly appreciated.

Note: We could decrypt our certificates which are encrypted using ciphers 3DES-SHA1 and AES256-SHA256. we need RC2-40-CBC for supporting out legacy products.

**Reply by t8m** [reply]
You should be able to use RC2_40_CBC with legacy provider loaded. The OpenSSL of course needs to be configured and build without `no-rc2` configure option.

**Reply by gururaghavendraprasad** [reply]
I have built the OpenSSL without no-rc2.
